libexec/apply_catalog.rb

```diff
@@ -48,7 +49,26 @@
     $LOAD_PATH << dir unless $LOAD_PATH.include?(dir)
   end
 
+  if (conn_info = args['_target'])
+    unless conn_info['type']
+      puts "Cannot collect facts for a remote target without knowing it's type."
+      exit 1
+    end
+
+    # TODO: validate_target
+    special_keys = %w[type debug uri]
+    connection = conn_info.reject { |k, _| special_keys.include?(k) }
+    device = OpenStruct.new(connection)
+    device.url = conn_info['uri']
+    device.provider = conn_info['type']
+    device.options = { debug: true } if conn_info['debug']
+    Puppet[:facts_terminus] = :network_device
+    Puppet[:certname] = device.name
+    Puppet::Util::NetworkDevice.init(device)
```

Comment:
@DavidS I need a way to do this with a transport.

Reply:
now I get it, bolt can't use the transport *alone*, because it needs to hook up the rest of puppet :man_facepalming: that took me way too long.

I believe you should be able to say
```
device = OpenStruct.new
device.url = connection
```

and have everything work out with the backwards compatibility layers we're building in for plain puppet.

I don't think that's a particular appealing long-term solution.

Would the following make sense?

```
transport = Puppet::ResourceApi::Transport.connect(conn_info['type'], connection)
wrapper = Puppet::ResourceApi::Transport::Wrapper.new(transport) # needs the wrapper to be modified, easy
Puppet::Util::NetworkDevice.current = wrapper # needs puppet to be modified (or use instance_variable_set)
```

Reply:
OK having chatted to @DavidS we could update NetworkDevice to be:
`Puppet::Util::NetworkDevice.init_transport(transport)` which then wraps the transport and sets current.

Reply:
`Puppet::Util::NetworkDevice.init_transport(Puppet::ResourceApi::Transport.connect(conn_info['type'], connection))`
,
`Puppet::Util::NetworkDevice.init_transport(conn_info['type'], connection)`
`Puppet::ResourceApi::Transport.set_device(conn_info['type'], connection)`


Code in puppet-resource_api instead of puppet will prevent this from being blocked on a puppet release.

Reply:
I'm happy to not have to wait for a puppet release and just force-feed from the outside

If we want, we can also monkey-patch an `init_transport()` function into `Puppet::Util::NetworkDevice` for puppet versions that don't carry it natively.

Reply:
We try not to monkey patch puppet from bolt. I think it's ok to require a new agent version on the proxy. In most cases it will be localhost and bolt should just use the agent that ships with it anyway.

Reply:
But I'd like to merge this in and release it without having to wait for puppet release if possible

Reply:
Bolt no longer vendors puppet we now just have a gem dependency on it

Reply:
In that case bolt needs to pull in the `puppet-resource_api` gem on its own anyways, and we can release the rsapi gem whenever we need it.